jupyter_client/manager.py

```diff
@@ -454,6 +454,80 @@ def is_alive(self):
             # we don't have a kernel
             return False
 
+    kernel_statistics = List(
+        Unicode(), config=True,
+        help="""Comma-separated list of support categories to gather relative to the active kernel.   
+                No statistics are gathered if option is not defined.  
+                Atm categories are: memory """
+    )
+
+    def kernel_stats(self, *stats):
+        """
+        Returns a dict containing functions that gather
+        different stats about the kernel.
+        Running the appropriate data gathering funcs according
+        to the args passed in.
+
+        Parameters
+        ----------
+
+        *stats: A list of strings representing the kinds of
+                statistics they are interested in collecting.
+
+        Usage
+        -----
+        >>> ks = km.kernel_stats("memory", "latency")
+        The function call above will return a dict with two key-value
+        pairs (assuming there are function to gather those stats).
+        The keys are exactly the same as the inputs.
+        >>> print ks["memory"]
+        Line above will print the memory usage of the kernel.
+        """
+        # `types` is a mapping of the desired stat type to the appropriate function
+        types = {
+            "memory": self._memory_usage
+        }
+
+        def get_stats(inputs):
+            result = {}
+            try:
+                for stat in inputs:
+                    try:
+                        func = types[stat]
+                    except KeyError:
+                        continue  # If the stat was not found in `types` dict
+                    else:
+                        result[stat] = func()
+                return result
+            except Exception as e:
+                self.log.debug(
+                    "error getting the stats of the kernel, {}".format(str(e))
+                )
+            finally:
+                return result  # Return what we were able to collect
+
+        # If an input was passed into the function
+        if stats and len(stats) > 0:
+            return get_stats(stats)
+        # If no input was passed but there is a default configuration
+        elif self.kernel_statistics and len(self.kernel_statistics) > 0:
```

Comment:
I don't think this `elif` is necessary.  The caller should be using `self.kernel_statistics` as the stats parameter to this method call, so if `stats` is None or empty, then it seems like this method should return the empty set.

I suppose the other approach, now that `kernel_statistics` is defined, is to not have a `stats` parameter at all and use `kernel_statistics` entirely in this method.  But having the `stats` parameter would allow callers to capture a subset of the supported categories that are specified in the configuration option.

Reply:
I thought about it some more, why would it be better to have this trait?
Wont it be better to just let the consumer of the function to decide what he wants to get,
without starting to handle some sort of configuration?
I see the reason for using a trait in a case of something that stays kind of static, for example the `notebook_dir` that you have linked the validation method to.
But this method is supposed to be flexible, we should not restrict the user in the choosing of the data he wishes to collect.
But by initializing the `kernel_statistics` we make it harder for the consumer of the function to quickly and easily change the data he gets.

> I don't think this elif is necessary. The caller should be using self.kernel_statistics as the stats parameter to this method call, so if stats is None or empty, then it seems like this method should return the empty set.

If we tell the user to call the function with the `kernel_statistics`, then it becomes much less flexible, because every little change we will have to change the parameter, instead of just changing the input to the function (also applies to new uses of the function).

What are the pros of having such trait? seems to me like its restricting in this case, whereas `*stats` is not.

Reply:
I'm sorry for the delayed response.  I guess I'm having difficulting understanding how the _consumer_ of this method is implemented.  I was under the impression these stats would be called, probably from the notebook _server_ code periodically.  Since that code is baked into the server, how does the end user control or influence what stats are gathered?  Nothing has been defined thus far that makes what is gathered quick and easy to change.  

I figured by exposing the "stat categories" via a configuration option, then _that_ would be how a given installation determines what stats are important and should be gathered.  In addition, this trait also serves as an enablement switch so that this functionality can be disabled (by default).  I don't follow why an installation would want to gather _different_ stats dynamically.  _What_ is gathered seems like a static thing to me.

At this point, I feel like things have gone sideways a bit and I apologize if you're feeling frustrated.  I believe the important things are that this functionality is disabled by default, what is gathered can be controlled by an admin, and all stats relative to a managed kernel are obtained in a single request.  I believe the current incantation addresses those points, but if you'd prefer to take this a different way, that's fine as well.

Perhaps @rgbkrk could lend some direction.

Reply:
@kevin-bates Its all good, I am all for having a debate over this.

Let me explain to you how I see this functionality being used, maybe then, you will see why I think that letting this function be dynamic is important.

1. Notebook server code that calls the function.
In this use case, there is a classic example that started this whole idea and that is the prometheus metric for memory usage of the kernel.
Image that we have different types of metrics that we want to let the prometheus agent to collect.
Those metrics are exposed through this function.
If you look at how prometheus collects different metrics, you will see that it is not happening all in one place.
different metrics are collected at different intervals, in different places of the notebook server code.
Thus, we have to let the code in the notebook server, the flexibility, the freedom to choose what he wishes to collect.

2. A frontend button that on a user click will return a visual representation of the kernel statistics.

This freedom of choice is given by the `*stats` parameter.
This is why I think that removing the `*stats` parameter will go against the main reason for all of this.
From what I have seen such flexibility is not wise to implement with the configuration of traitlets, which are supposed to be static as soon as you start the server.

Are we on the same page here? or is there something still unclear about my perspective?

Reply:
> From what I have seen such flexibility is not wise to implement with the configuration of traitlets, which are supposed to be static as soon as you start the server.

As an operator, I definitely prefer config traitlets that stay the same after starting the server.

I haven't read back into this PR much, but I'd say that we definitely want `jupyter_client` to behave more and more like a library (accepting options as objects passed in). Only applications should use traitlets for config. cc @minrk

Reply:
@rgbkrk Cool I agree.
For a second, lets say I remove the traitlet from the code.
We are left with the implmenetation of the method `kernel_stats` that accepts `*stats` as prameter, and returns the wanted gathered data types as long as it is possible.

Now, my last concern is that, it might be better to not complicate things, and just leave the `memory_usage` method as is, without encapsulating it inside the `kernel_stats` method.

But the thing is, I do see the positive side of having one method that encapsulates all of the gathering methods inside of it.
It creates a defined one way of getting the statistics that you want from the kernel, a json that you know what to expect inside.
Also it will save lines of code.
But it feels robust, which is good, but it comes in place of simplicity in my opinion.
There is some hidden mechanism inside the function `kernel_stats` whereas in plain calls directly to the `memory_usage` method there aren't.

@rgbkrk , @kevin-bates , thoughts?